RecoLocalMuon/CSCSegment/src/CSCSegAlgoTC.cc

```diff
@@ -792,214 +533,82 @@ void CSCSegAlgoTC::pruneTheSegments(const ChamberHitContainer& rechitsInChamber)
   segmentSort();
   
   // Select best quality segments, requiring hits are assigned to just one segment
-  // Because I want to erase the bad segments, the iterator must be incremented
+  // Want to erase the bad segments, so the iterator must be incremented
   // inside the loop, and only when the erase is not called
 
-  std::vector<ChamberHitContainer>::iterator is;
-  std::vector<double>::iterator ichi = chi2s.begin();
-  std::vector<AlgebraicSymMatrix>::iterator iErr = errors.begin();
-  std::vector<LocalPoint>::iterator iOrig = origins.begin();
-  std::vector<LocalVector>::iterator iDir = directions.begin();
+  std::vector<CSCSegFit*>::iterator is;
 
   for (is = candidates.begin();  is !=  candidates.end(); ) {
     
-    bool goodSegment = isSegmentGood(is, ichi, rechitsInChamber, used);
+    bool goodSegment = isSegmentGood(is, rechitsInChamber, used);
     
     if (goodSegment) {
       LogDebug("CSC") << "Accepting segment: ";
-      
       flagHitsAsUsed(is, rechitsInChamber, used);
       ++is;
-      ++ichi;
-      ++iErr;
-      ++iOrig;
-      ++iDir;
     }
     else {
       LogDebug("CSC") << "Rejecting segment: ";
-      is = candidates.erase(is);
-      ichi = chi2s.erase(ichi);
-      iErr = errors.erase(iErr);
-      iOrig = origins.erase(iOrig);
-      iDir = directions.erase(iDir);
+      delete *is; // delete the CSCSegFit* 
+      is = candidates.erase(is); // erase the element in container
     }
   }
 }
 
 void CSCSegAlgoTC::segmentSort() {
   
-  // The segment collection is sorted according chi2/#hits 
+  // The segment collection is sorted according to e.g. chi2/#hits 
   
-  for(unsigned int i=0; i<candidates.size()-1; i++) {
-    for(unsigned int j=i+1; j<candidates.size(); j++) {
+  for(size_t i=0; i<candidates.size()-1; ++i) {
+    for(size_t j=i+1; j<candidates.size(); ++j) {
       
-      ChamberHitContainer s1 = candidates[i];
-      ChamberHitContainer s2 = candidates[j];
-      if (i == j)
-        continue;
+      size_t ni = (candidates[i]->hits()).size();
```

Comment:
isn't there a ``` candidates[i]->nhits() ``` ?
it seems wasteful to create a copy of hit pointers to just get the size out of it, and do it in the double loop